### DISCUSSION TITLE
UI ideas

### DISCUSSION THREAD

**Discussion by max-baz** [isamert/scli#167] [Ideas]
I like the current UI, but as things can always be improved, I was curious to see if you are on a lookout for ideas how to optimize the look and feel 🙂 

I really like the UI of thelounge.chat - a screenshot is below, of a public channel, so hopefully nobody will get offended 😄 

![image](https://user-images.githubusercontent.com/1177900/151857072-a3400def-ce2f-4eee-aba2-31adc132cf68.png)

Things to note:

- We can still show usernames for every message (as with `show_names=true`) and yet double the amount of messages shown on a single page, as username no longer takes a line, but is shown on a side to the message. Real estate is expensive!
- We can also see time of messages, which I realized we don't have at all in the current UI
- A "Today" separation takes a line, but allows to save horizontal space, as otherwise we'd have to write a date in front of every message. If I scroll up in the history, instead of "Today" it would say "Yesterday" or "29 January 2022"
- The screenshot above doesn't contain sent/delivered/read icons, but we should obviously preserve them.

What do you think overall? Interesting? Dislike? Too much work to change?

**Comment by exquo** [comment]
Ideas always appreciated!

This type of layout seems to be pretty standard for the IRC clients. Would be nice to have an option to reproduce (or approximate) it.

- Contact name on the same line as the message text
	Shouldn't be difficult to implement. One thing to note: on the screenshot the names are tab-aligned. Implementing this would be less straightforward. With "theirs-ours", left-right alignment we have by default (without `--one-sided`) it also won't look as neat.
- Show messages' times
	Yes, currently it's only found in the MessageInfo window (with `i`). It's not included in the chat view to save space. At least in my case, the exact times of most of the message I get (or send) is not important, or is (implicitly) known - like in a real-time back-and-forth chat. And when needed, one can check the MessageInfo.
- Reactions are also currently taking up a whole line to themselves. They could also be placed on the same line as the message (although that increase a chance of them going unnoticed there)

Personally, I like a somewhat spacious, vertically-separated layout, so that the messages are not packed too tightly against each other. But for those whose tastes run in the opposite direction of mine (which are probably the majority of users :), perhaps we can have a 'compact' / 'single-line' mode  setting. With it, all of the messages' components will be printed in a single line, like
	
	>> 01:23 Alice: Hello there!   [👋]   
	                                               Hi!  12:34 ✓✓

Where the emoji in square brackets is the reaction. Adding a quoted message to the same line will probably look unwieldy, so best keep them multi-line. Also, some messages have explicit newlines.

With the above implementations in place, we can have an "IRC-like" mode (except for its three-column aligned layout: time - sender - message text), with the combination of `--one-sided --show-names --show-time --inline-message-mode` (the last two don't exist yet). 

I think it should be doable. In fact, changing a single separator from `\n` to ` ` (space) in [this line](https://github.com/isamert/scli/blob/585122c4ff9aa06597a1e6bf95ce0c6b18e7aed5/scli#L3037) already makes it look like a "single-line" mode. With adding a method for a date-time markup, it should be most of the way there. As often happens though, the devil might be uncovered in the details :).

**Reply by max-baz** [reply]
> - Contact name on the same line as the message text
> Shouldn't be difficult to implement. One thing to note: on the screenshot the names are tab-aligned. Implementing this would be less straightforward. 

Not straightforward but worst case just computing and padding the names manually with spaces, given a known target width, shouldn't be too difficult 🙂  I do believe it can increase readability, so lets at least try to explore - I am happy to help 😉 

> With "theirs-ours", left-right alignment we have by default (without --one-sided) it also won't look as neat.

I think in a group it could be quite useful, no? Of course we only pad the left side:

```
>> 01:23     Alice: Hello there!   [👋 👋]   
>> 01:24       Bob: Greetings!   [👋]   
                                                           Hi!  12:34 ✓✓
```

> At least in my case, the exact times of most of the message I get (or send) is not important, or is (implicitly) known - like in a real-time back-and-forth chat. 

That is true in many cases, but quite useful when you are catching up on backlog, if you have chatty friends while you are asleep 😁 

Also, a general point, with the modern wide screen we usually can fit more horizontally than vertically, so I am not too worried about having to use ~10 extra chars for time... As you can see in this thread I'm quite more interested to save vertical space 😁 

> Reactions are also currently taking up a whole line to themselves. They could also be placed on the same line as the message (although that increase a chance of them going unnoticed there)

That's an interesting idea! I initially thought that with multi-line messages this would look bad (especially if lots of other emoji symbols are involved, e.g. with the currently popular wordle), but it doesn't look bad at all!

```
>> 01:23     Alice: Wordle 227 6/6
                    ⬜🟨⬜⬜⬜
                    ⬜⬜🟩⬜⬜
                    ⬜⬜🟩⬜⬜
                    🟨🟨🟩⬜⬜
                    ⬜⬜🟩🟩🟨
                    🟩🟩🟩🟩🟩  [👍]
>> 01:24     Alice: Hey Bob!
                    Show us what you got! [👍 😆]
>> 01:33       Bob: Wordle 227 3/6
                    🟨⬜⬜🟨⬜  
                    🟩🟨⬜🟩🟩
                    🟩🟩🟩🟩🟩
                    I'm on fire 😎  [👍 🤯]
```

> Adding a quoted message to the same line will probably look unwieldy, so best keep them multi-line. 

Absolutely agree

> I think it should be doable. In fact, changing a single separator from \n to (space) in this line already makes it look like a "single-line" mode.

Haha, indeed! 😁 To me it also very quickly proves that tab-aligning names would be very important, otherwise my eyes get tired jumping left and right trying to identify where messages begin...

> `-one-sided --show-names --show-time --inline-message-mode`

The only downside I can think of is that it can be quite complex to keep all possible UI configurations stable and without bugs... as demonstrated by my recent PR where I was discovering on the go which other combinations I forgot to account for 😅

**Reply by exquo** [reply]
> Not straightforward but worst case just computing and padding the names manually with spaces, given a known target width, shouldn't be too difficult 

The complication I was thinking about was determining and adjusting the target length itself, on the fly, given the available screen space and the names in the view (and adjacent). But that might be over-complicating it. As a first step, we can just have a hard-coded length.


> lets at least try to explore - I am happy to help 😉

I'll look at this again soon-ish. It might indeed turn out to be not hard. Contributions welcome, of course!


> I think in a group it could be quite useful, no? Of course we only pad the left side

Yep, I can see the utility of this. We can always pad the names when they are shown on the left.


> with the modern wide screen we usually can fit more horizontally than vertically

True, vertical space is likely more valuable for most users.



> I initially thought that with multi-line messages this would look bad (especially if lots of other emoji symbols are involved

We could also check whether a message itself contains newlines, and place reactions on a new line in that case.

Bonus points: adding a `--wordle-mode` option :) (whatever that'd look like)


> tab-aligning names would be very important, otherwise my eyes get tired jumping left and right trying to identify where messages begin...

Makes sense. However, the names are shown in bold, unlike messages. And we could add other separating markers.


> `-one-sided --show-names --show-time --inline-message-mode`
	The only downside I can think of is that it can be quite complex to keep all possible UI configurations stable and without bugs... 

Exactly what I was thinking! We don't yet have a test suite, so unfortunately the permutations have to be tried manually.

**Comment by exquo** [comment]
I have added options `--show-message-time` and `--show-inline` in the [`feature/167-ui-inline` branch](https://github.com/isamert/scli/tree/feature/167-ui-inline). Running with
	
	scli  --show-names --show-message-time --show-inline --one-sided  

results in an interface similar to the one in the screenshot above.

Some differences: reactions are still shown on a separate line; the `--show-time` option allows to specify a custom time format, but there is no "date line".

Let me know if there are any comments!

**Reply by max-baz** [reply]
That's quite interesting! Thanks for working on it!

The most important thing that negatively stood out to me is how multi-line messages are currently displayed:

```
>> 21:28  James Smith Some very very very long 
   line
✓✓ 21:35           Me huh
```

I'd *much* prefer for the message contents to not be blended with author names, e.g. like so:

```
>> 21:28  James Smith Some very very very long 
                      line
✓✓ 21:35           Me huh
```

Even better, if we can have some visual separation for the two columns, people and messages, something that is easier for the eye to "catch on" than just the bold font of the author names. Perhaps like this?

```
>> 21:28  James Smith | Some very very very long 
                      | line
✓✓ 21:35           Me | huh
```

This way, I feel like I can more easily glance at the messages alone, I make less mental effort to visually distinguish where do messages begin.

What do you think?

**Reply by exquo** [reply]
Thanks for the feedback! I agree, it'll look better with a separate vertical column for the message's text. I'll add it.

Although for messages that are longer, the current layout is also readable:

<code>>> 21:28  <b>James Smith</b> Sed ut perspiciātis unde   
	 omnis iste nātus error sit voluptātem accūsantium          
	 dolōremque laudantium, tōtam rem aperiam,                  
	 eaque ipsa quae ab illō inventōre vēritātis                
	 et quasi architectō beātae vītae dicta sunt                
	 explicābō.                                                 
✓✓ 21:35           <b>Me</b> huh                              
</code>


and it uses up less space than the columns-based alternative. But most of the potential users of the `inline` layout are likely to have wide screens with a lot of horizontal space in the terminal windows.

**Reply by max-baz** [reply]
Yeah! Regardless of a screen size, I feel like I would definitely at least keep `show-inline`, I like it :slightly_smiling_face: Maybe the inline quote (and attached, etc?) could be made in italic font, or using dim-white color, again to visually separate the message from the quote? Just some ideas :thinking: 

Have you considered showing the time before `>>` and `✓✓`? Not to confuse it with part of the message, especially without showing author name.

I don't disagree that the current layout is readable, but for some reason I still feel like a column-based UI would still be easier to spot at a glance, skim to the next message, etc... On a smaller screens, I would maybe omit the author names and at least try to include `show-time` and with it I'd already prefer to have column layout... :thinking:

**Reply by exquo** [reply]
With the added options the number of possible combinations becomes quite large. I have considered allowing instead to specify the included items and their layout through a format string to `strftime`. E.g. `%(date) %(status) %(name!column:12) \n %(message_text)` etc.
But creating a new domain-specific language for this seems like an overshoot. And allowing to run a user-supplied python code might be a problem.

So for the options and their combinations I'll have to pick some best "middle-of-the-road" results that would be acceptable to most of the people.
Thanks for providing the suggestions! I'll try to figure out which would work best.

**Comment by exquo** [comment]
I've pushed some more changes to [feature/167-ui-inline](https://github.com/isamert/scli/tree/feature/167-ui-inline).
 
The quoted message text is now in italics (thanks again for the suggestion!). Would be nice to have a better visual separation method, but in a terminal app the options are more limited than in a typical GUI.. Using color is a good idea, but it should only be applied when `--color` option is given; by default, we should stick to the monochrome palette for maximum compatibility across different setups. 
The quoted text is still placed on a separate line from the main message, even in the `inline` mode; this should further help with establishing a visual break. Finally, the separator characters (`|` and `|>` in case of inline quote) are still there too.

> Have you considered showing the time before `>>` and `✓✓`? Not to confuse it with part of the message, especially without showing author name.

That would create a better separation in that case, but I think in general it's better to keep the existing ordering: 
- since the time column is always the same size (it's also vertically aligned), that makes it easier to separate it from the message
- most people who use the inline mode, probably would also use the `--show-names` (to achieve the results similar to the reference screenshot above). Since the names are in bold font, they add a visual break.
- those same users are also likely to use `--one-sided`. Having the `>>` or `✓✓` as the first marks on a line makes it visually clear who the sender of each message is. 

WRT columns: I agree, the column-based layout is better for inline mode. It's now a default when `--show-inline` is used. But I've also added back the wraped-text layout as an option that can be turned on with `--show-inline=wrap`.

Let me know if that sounds (and looks) reasonable!

**Comment by max-baz** [comment]
I like it a lot, and agree with everything you said :+1: I do not quite understand one part, maybe you can help me here:

>  The quoted text is still placed on a separate line from the main message, even in the inline mode; this should further help with establishing a visual break. Finally, the separator characters (| and |> in case of inline quote) are still there too.

In other words, quote placement is equal both with and without `--show-inline`, correct? I can't find an example where they would take different amount of space, or otherwise differ in anything except the separator characters. In so, why don't we keep just one way to represent the quote, by using `|` as on master that we already know and love? 

```
✓  Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam
   erat volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo
   consequat.
✓  | Maxim Baz
   | Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed [...]
   Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam
   erat volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo
   consequat.
✓  one
   two
   three
   four
✓  | Maxim Baz
   | one two three four
   five
   six
   seven
```

vs

```
✓  Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam
   erat volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo
   consequat.
✓  |> Maxim Baz
   Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed [...] |
   Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam
   erat volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo
   consequat.
✓  one
   two
   three
   four
✓  |> Maxim Baz
   one two three four |
   five
   six
   seven
```

**Reply by exquo** [reply]
With `--show-inline`, the quoted name and message text were intended be placed on the same line. In the last commit they aren't, but that's a regression. (So many display options, too many possible combinations to test!:) Thanks for noticing, I have fixed that in a new commit.

A line break is included between the quoted message and the reply (i.e. the main message text), even with the `inline` mode. That's what I've meant in the passage quoted above.

The `|>` quote "indicator" was introduced for the `inline` mode instead of just the `|` to make the quotation stand out visually better, since there could be other elements on the same line (time, sender's name).

**Reply by max-baz** [reply]
Ah that makes sense! I can see that in the regular mode, the quote can take up to two lines (name + 1 line of content), after which it's clipped, and in the inline mode it is also clipped but it "slips" a little bit on the second line, so the max number of lines it can take is still two. Is there some specific math going on, or is it just a coincidence?

```
✓  20:02           Me Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam
                      nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam erat
                      volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation
                      ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo consequat.
✓  20:02           Me |> Maxim Baz Lorem ipsum dolor sit amet, consectetuer adipiscing elit,
                      sed [...] |
                      Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam
                      nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam erat
                      volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation
                      ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo consequat.
✓  20:03           Me one
                      two
                      three
                      four
✓  20:03           Me |> Maxim Baz one two three four |
                      five
                      six
                      seven
```

```
✓  20:02 Me
   Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy nibh euismod
   tincidunt ut laoreet dolore magna aliquam erat volutpat. Ut wisi enim ad minim veniam,
   quis nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo
   consequat.
✓  20:02 Me
   | Maxim Baz
   | Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed [...]
   Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy nibh euismod
   tincidunt ut laoreet dolore magna aliquam erat volutpat. Ut wisi enim ad minim veniam,
   quis nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo
   consequat.
✓  20:03 Me
   one
   two
   three
   four
✓  20:03 Me
   | Maxim Baz
   | one two three four
   five
   six
   seven
```

I'm asking because, if in both display types, the quote is allowed to take at most two lines, maybe it will still make it easier for you to keep just one way of representing the quote?

I don't have any particular preference for `|` vs `|>` for inline mode, personally I think both are equally fine.

Whether we use `|` or `|>`, given that there is always a newline between a quote and a reply, maybe it doesn't make sense to end the quote with `|` in the inline mode? I'd rather have the second line of quote to also start with the indicator (so, assuming we keep the new indicator and assuming we have two lines, I'd prefer the below):

```
✓  20:02           Me Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam
                      nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam erat
                      volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation
                      ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo consequat.
✓  20:02           Me |> Maxim Baz Lorem ipsum dolor sit amet, consectetuer adipiscing elit,
                      |> sed [...]
                      Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam
                      nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam erat
                      volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation
                      ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo consequat.
✓  20:03           Me one
                      two
                      three
                      four
✓  20:03           Me |> Maxim Baz one two three four
                      five
                      six
                      seven
```

**Reply by exquo** [reply]
The quoted text is currently clipped at 70 symbols. That is, it is limited in *length*, rather than in the number of lines it can use. Number 70 is also somewhat arbitrarily, of course. In normal mode it should be able to fit on a single line of a typical terminal window. However, when run with `--show-inline` it might not - it depends on the width of a given terminal window and the amount of other stuff on the same line: time string, sender's name, etc.

Limiting the quote's text to a single line is unfortunately not straightforward, because the quote is only one part of a markup string supplied to an `urwid.Text` widget (along with the reply, the list of attachments, etc). `Text` by itself does not distinguish between different types of markup elements. The best it can do is to clip the *whole* Text element to a single line. There are ways around this, but currently for simplicity's sake I just cut it at 70 chars. 
Something similar goes for adding some character (e.g. `|`) at the start of every line: this would require writing a custom `TextLayout` class for urwid (which I might get around to doing one day), but the stock `urwid.Text` does not do have this capability.

> given that there is always a newline between a quote and a reply, maybe it doesn't make sense to end the quote with `|` in the inline mode? 

That's a good point, I might remove the trailing `|` there. Thanks!

**Reply by max-baz** [reply]
I see, thanks for clarifying!

**Comment by max-baz** [comment]
I seem to have a lot of friends with long names :sweat_smile:  For me `SENDER_NAME_COL_WIDTH = 12` results in almost everyone's name being wrapped :sweat_smile: The smallest value that fits everyone is `17`.

It's not that I think one number is more correct than another, I totally realize it's hard to come up with one perfect hardcoded number, and I do like that the column is "fixed" across all chats, so the layout doesnt become wider and narrower all the time as I switch between people. However, I don't like that when the author name wraps, it basically results in 50% of vertical space being lost, as now there's empty line between each message. This entire discussion, after all, is about conserving vertical space at the expense of horizontal one :grin: 

```
>> 00:11 John       Hi
         Smith
>> 00:11 John       How are you?
         Smith
>> 00:14 John       Are you busy?
         Smith
```

What do you think is the best way to handle it?

**Reply by exquo** [reply]
Name column length: that's another regression (thanks for bringing this one to the attention as well!). The names were supposed to be *clipped* at N symbols (N=12 chosen somewhat arbitrarily, of course). Wrapping longer names onto a second line certainly is not a good idea for the `inline` mode. I've pushed a commit to resolve that also.

In general, it would be nice to allow specifying the sender's name column's width as a parameter (i.e. to not hardcode the N), but it's not straightforward to use one of the existing options for this: the names are only shown in the fixed-width column when `--show-inline=columns`, `--show-names` and `--one-sided` are used. So making the width an optional parameter to e.g. `--show-name=12` would not make sense generally. And creating yet another display-related option (e.g. `--name-column-width`) does not seem to be sufficiently necessary. So having some reasonable hard-coded width looks good enough. Let me know if you think the current value (12) might be too small (or large). Of course there's no number that would suit everyone perfectly, but just something convenient for most people.

**Reply by max-baz** [reply]
I agree entirely, and besides, as a user I don't _want_ to make the decision about the names column length. I think clipping with some sensible default is a valid strategy, but I find it really difficult to come up with arguments for or against `12` :grin: What do you think about going through the contact list on startup, ignoring groups, and using the longest name as the limit? It is still possible that you don't have the `Mr Excessively Long Name` in your immediate contacts, but they are part of a group you are in, in which case their name would be clipped.

**Reply by exquo** [reply]
> What do you think about going through the contact list on startup, ignoring groups, and using the longest name as the limit? 

Something like that is possible, but probably not really necessary: in one-to-one chats, the name of the contact is not needed on every message (after all, it's known whom a given chat is with); it would really suffice to just have "me" and "them" indicators. And 12 symbols would be enough to fit a meaningful amount of the name (in many cases - fit completely). 
And in groups, where `show-names` is actually more important, there might well be some `Mr Excessively Long Name`s :) So clipping at some reasonable length is better than having an extra wide column taking up all the horizontal space.

**Reply by max-baz** [reply]
I actually agree with you, after using `scli --show-message-time --show-inline --show-names` for the last couple of days as my daily driver, I tend to think that in 1:1 chats the name *is* not needed... as you correctly say, I always know whom a given chat is with. And in the groups, the names are anyway shown even without `--show-names` today.

Even in the original screenshot in this thread, the names are shown there exactly because it's a group conversation! 

So a big question is now, do we really need `show-names`?

A side-note: in a group conversation, currently `show-names` does make a difference: everyone else's names are always shown, but `Me` is shown only with `show-names`, without this flag, the "grid" is broken as the "rows" with my messages lack the "name" column.

Whether or not we keep `show-names`, would it make sense to always show `Me` in the group conversations, just like we always show everyone else's name?

**Reply by exquo** [reply]
> So a big question is now, do we really need `show-names`?

It's a "pre-existing" option in `scli`, not introduced by the changes in this feature branch. So it needs to stay for backwards compatibility. Somebody probably finds it useful, especially in group chats with `one-sided`. And now, in combination with `inline` and `one-sided`, it allows to have a 3-column layout even in one-to-one chats, which some people might prefer for uniformity.

> shown only with `show-names`, without this flag, the "grid" is broken as the "rows" with my messages lack the "name" column.

Indeed, this does not look right. Without `inline` it might not have been necessary to display `Me` in place of the sender's name: when `one-sided` was not used, the sent messages were printed on the right side; and even with `one-sided` it looked okay (it was possible to distinguish between the sent and the received messages). But now that it breaks the grid it should be fixed. Thanks for spotting it! I've added a fix that automatically includes a `Me`-column when both `show-inline` and `one-sided` are on, even when `show-names` has not been specified manually.

**Reply by max-baz** [reply]
Makes sense, and thanks, it looks good now in the group channel! :slightly_smiling_face:

**Comment by exquo** [comment]
Another change pushed to the `inline` mode: now the reactions are also shown on the same line.

**Reply by max-baz** [reply]
Exciting! 

On the one hand, I'm loving that it saves some space - in one of my groups in particular, there are a ton of reactions going, so this change literally allows me to fit twice as much content on the screen in that channel! 

<details>
<summary>snip</summary>

![image](https://github.com/isamert/scli/assets/1177900/f184cb6c-66fe-4703-8a74-499dbf5ab087)

</details>

On the other hand, especially in a "normal" chat with only occasional reactions, because it is aligned by the right side, there is so much space between the message and the reactions, it's really difficult to match a specific message to a reaction - I can see somewhere on the right side that _some_ message had a reaction, but I almost have to take a ruler to find out which message it corresponds to...

I don't think there's particular need to have the reactions be aligned in a column, like names and message content is... What do you think about adding reactions to the end of the message? In the same format, just not as a column aligned to the right, but appended to the end of the message - have you tried it already, perhaps?

On a side-note, it may be a bit too early to report this kind of stuff, but there's a crash if you try to add, change and remove reaction on a message (I tried on a message in "Note to Self" channel):

<details>
<summary>snip</summary>

```
  File "/home/maximbaz/scli/./scli", line 4038, in _reaction_picked
    action_request.send_reaction(self._msg, emoji)
  File "/home/maximbaz/scli/./scli", line 4482, in send_reaction
    self._daemon.send_reaction(
  File "/home/maximbaz/scli/./scli", line 1064, in send_reaction
    self.callbacks['sending_reaction'](envelope)
  File "/home/maximbaz/scli/./scli", line 4855, in _on_sending_reaction
    self._add_reaction(envelope)
  File "/home/maximbaz/scli/./scli", line 4989, in _add_reaction
    msg_w.update_reactions_w()
  File "/home/maximbaz/scli/./scli", line 3513, in update_reactions_w
    self._reactions_w.update(emojis_markup)
  File "/home/maximbaz/scli/./scli", line 3177, in update
    if align == 'right' else
       ^^^^^
NameError: name 'align' is not defined
```

</details>

**Reply by exquo** [reply]
> What do you think about adding reactions to the end of the message? In the same format, just not as a column aligned to the right, but appended to the end of the message

That's the most straightforward approach; the problem with it is that it makes it hard to distinguish between a message and a reaction to it. I have put together some changes that accomplish this in 0e2d84e2, just to check what that would look like.

Having reaction in a separate column seems like a better idea, as it gives a clearer visual separation. But I can see how with a wide terminal window the gap between the reactions and the message text becomes too wide. One solution to this is to bring the columns closer together. I have made some changes that do that in the latest commit. Let me know whether that works for your case. Also, you can try to play with the number in [this line](https://github.com/isamert/scli/blob/feature/167-ui-inline/scli#L3534) to see if some value of the reactions column's weight works better than the current one.


> On a side-note, it may be a bit too early to report this kind of stuff, but there's a crash if you try to add, change and remove reaction on a message (I tried on a message in "Note to Self" channel):

Not too early at all! Thanks, should be fixed now.

**Reply by max-baz** [reply]
I tried using both versions for a significant amount of time, I think the https://github.com/isamert/scli/commit/0e2d84e2e0d93503e11432e1bbfe1c76e6998ca0 looks a lot better, even compared with a reduced column width of the alternative idea. At the same time I have to admit that for my taste, the reactions shown on a separate line (as in the release) are prettier than any of the two inline options... Probably because of what you said is true, that when reactions are on the same line, they are somehow mentally mixed with the message itself.

I was considering if something like this could be possible, but it has a lot of challenges, such as when reactions don't fit in that space, or without `--show-time`... it doesn't feel like it would be feasible to make in a generic way, would you agree?

```
✓  20:02 Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy nibh euismod tincidunt ut laoreet dolore magna
╰╴👍 🙌  aliquam erat volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex
         ea commodo consequat. 

```

**Reply by exquo** [reply]
Thanks for the feedback!

I think it should be doable to implement something like your example above. It does look like a very nice way to show reactions for this case: if the second column exists (containing time or sender's name), and if the message is long enough to span several lines.

But, like you said, if the `time` (or `name`) column is not turned on, this look will not be achieved. Also, most messages are likely to be short enough to fit on a single line, especially on wide screens, where people are more likely to use `inline` mode. In that case reactions will take up an extra line - not great for the `inline` mode. And if the message is actually long enough to span more than one line, then there is no ambiguity with visually matching the messages to reactions (discussed above), so the column layout should work okay there.

Reactions appended to the end of the message text might be hard to notice at all. A message text itself might well contain the very same emoji; and the message widget does not get noticeably transformed when a reaction arrives and an emoji is appended this way. 

So, it seems like the column layout is the best option for a general case. Maybe there are ways to make it more visually appealing; if I think of something (or if there are any suggestions) I'll be sure to make improvements!
